### DISCUSSION TITLE
Map keyboard key to mouse movemennt

### DISCUSSION THREAD

**Discussion by dnikic** [sezanzeb/input-remapper#63] [Ideas]
I would like to use this app to move my cursor with vim binding + meta-key.
Joystick to mouse movement is implemented, but i didn't find key to mouse movement, is that present or should I try to write something up a send a Pull Request?

**Comment by sezanzeb** [comment]
Something like that is not present.

Sure, contributions are welcome! I have tried my best to make the code very documented and clear, but I guess it still got quite complicated here and there. If something is confusing, feel free to ask.

Using joysticks to move the mouse is implemented via:
- `def should_map_as_btn`, which looks at the configuration (`mapping.get('gamepad.joystick.left_purpose')`) to figure out if an event may be used to map to a button at all. If a joystick is used as mouse it returns False
- `def run` of `event_producer.py`, which keeps moving the cursor as long as the joystick is not released. A held down button/joystick will not report new events, so some sort of loop was needed for that. The debouncing stuff in there has nothing to do with it.
- `def is_handled` and `def nofify` also of `event_producer.py`, which is used to consume joystick events. In this case it would instead accept the events that map to cursor movements.

I hope the existing architecture can support that feature without too much struggle. What do you think of using keywords like `mouse_left`, `mouse_up`, etc. for the mapping? But it would not be possible to control the mouse speed by just using keywords.

Please also write tests for your changes

**Reply by sezanzeb** [reply]
why do you need that btw, don't you have a mouse on the computer?

**Reply by dnikic** [reply]
Thanks for the fast reply and detailed answer!
Ill clone the repo and look into the code as soon as i catch some free time. 

> What do you think of using keywords like mouse_left, mouse_up, etc. for the mapping? 

Sounds good, but it would be event better if an user could chose the cursor movement speed or move distance.

**Reply by dnikic** [reply]
> why do you need that btw, don't you have a mouse on the computer?

Developing a hand injure from using bad mice got me interested in ergonomics while programming.

**Reply by sezanzeb** [reply]
I guess I would prefer to display a slider if mouse_* mappings are used, basically similar to the existing sliders.

That can be also added later on in a separate pull request, whatever fits you

**Reply by sezanzeb** [reply]
I think `should_map_as_btn` doesn't need to be modified, only `is_handled`

**Reply by sojusnik** [reply]
> > why do you need that btw, don't you have a mouse on the computer?
> 
> Developing a hand injure from using bad mice got me interested in ergonomics while programming.

Do you know [keynav](https://github.com/jordansissel/keynav)? Could be handy in your situation.

**Reply by dnikic** [reply]
> Do you know keynav? Could be handy in your situation.
Thanks, ill look it up.

**Comment by sezanzeb** [comment]
I just realized it would be fairly easy to add via macros. You can try it on the https://github.com/sezanzeb/key-mapper/tree/write-macro branch. Took 20 minutes or something to implement

![image](https://user-images.githubusercontent.com/28510156/111826076-8a5be300-88e8-11eb-8260-0fcf1190312d.png)

`write` could be used to inject all sorts of events via the keyboard. scrolling, gamepads, whatever. Maybe I'll rename it to `e` as in "event". `w` is already taken by "wait". I abbreviate all of them due to the space constraints. It would be possible to make writing that stuff easier via constants (see https://github.com/torvalds/linux/blob/master/include/uapi/linux/input-event-codes.h)

what do you think?

`mouse(up, 4)` might be a handy readable shortcut that could be implemented.

And via that method it is possible to control the mouse speed as well

**Reply by sezanzeb** [reply]
done:

![image](https://user-images.githubusercontent.com/28510156/111843476-93a57980-8901-11eb-9705-bb2fc71ae2bb.png)

`e(EV_REL, REL_X, 10)` also works

now only tests are missing. sorry if you were looking forward to making it yourself. I just couldn't stop

**Reply by sezanzeb** [reply]
tests done, pushed to main

**Reply by dnikic** [reply]
This is AWESOME!
Congratulations for being so active and fast, respect!
The only missing feature i saw is having mouse() in the mapping drop-down.

Is it possible to toggle activation of a mapping layer on click instead of pressing a key combo?
Let me illustrate with an example.
I press CapsLock and my h,j,k,l become mouse movement buttons, g,f,d become mouse clicks, and s,a become scroll up and scroll down. When i press CapsLock again, my keys become regular letters.

**Reply by sezanzeb** [reply]
Thanks

a single press on capslock to switch the mapping of h, instead of mapping capslock + h?

that is not possible at the moment. switching presets quickly will be rather difficult, currently it stops the injection and starts a new one when doing that. Assigning shortcuts to presets is also not implemented, the injection daemon doesn't support any shortcuts yet. So it would be a relatively hard to implement

**Reply by dnikic** [reply]
Even without the caps lock preset switching, this is a much better and ergonomic solution than the default mouse keys feature in Gnome.
In my opinion, this app with a mouse preset should be a default app in an OS.
Great job!

**Reply by sezanzeb** [reply]
> The only missing feature i saw is having mouse() in the mapping drop-down.

it's there now

**Reply by dnikic** [reply]
Looking great!
Ive found that scroll down doesn't seem to e working (at least on my Ubuntu 20.04 system).

Would it be possible to apply a current preset and reset to default?
If not it this behavior may be accomplished by checking autoload and setting a OS binding to running and kill the key-mapper app.

**Reply by sezanzeb** [reply]
> Would it be possible to apply a current preset and reset to default?

> OS binding

??

**Reply by dnikic** [reply]
You have already answered my first question, with 

>  the injection daemon doesn't support any shortcuts yet.

By OS binding i meant setting up a hotkey in Gnome or other DE.

**Comment by cutbypham** [comment]
- Video demo: https://www.youtube.com/watch?v=pKtIrQsbKnQ
- Hold `Tab` 
+ `h` `j` `k` `l` moving cursor `left` `down` `up` `right` (vim) || hold `tab + space` + hjkl speed is slow down
+ `w` button left, `e` button middle, `r` button right
+ `;` scroll down, `p` scroll up, `i` scroll left, `o` scroll right
> I use qwerty layout
 - With this basic ideas you can use input remapper or the software you can think can do with action
- Input remapper config: https://github.com/thuanowa/dotfiles/blob/main/input-remapper/presets/SEMICO%20%20%20USB%20Gaming%20Keyboard%20/thuan_fly_key.json
